src/lib/array.jl

```diff
@@ -8,8 +8,26 @@ using Distributed: pmap, AbstractWorkerPool
 
 @adjoint copy(x::AbstractArray) = copy(x), ȳ -> (ȳ,)
 
-@adjoint collect(x::Tuple) = collect(x), dy -> (Tuple(dy),)
-@adjoint collect(x::AbstractArray) = collect(x), dy -> (dy,)
+@adjoint function collect(x::Tuple)
+  collect_tuple_pullback(dy) = (Tuple(dy),) 
+  collect(x), collect_tuple_pullback
+end
+
+@adjoint function collect(x::NamedTuple{names}) where names
+  collect_namedtuple_pullback(dy) = (NamedTuple{names}(Tuple(dy)),) 
+  collect(x), collect_namedtuple_pullback
+end
+
+@adjoint function collect(x::AbstractArray)
+  collect_array_pullback(dy) = (dy,)
+  collect(x), collect_array_pullback
+end
+
+@adjoint function collect(d::AbstractDict)
```

Comment:
I just noticed `reconstruct_if_dict` always returns a `Dict`. In that light, tightening this signature to 
```suggestion
@adjoint function collect(d::Dict)
```
feels like a good idea. No need to think about what would happen with IdDicts or more exotic types.